[CodeGen][AArch64] ptrauth intrinsic to safely construct relative ptr

AbhayKanhere opened the pull request:
ptrauth intrinsic to safely construct relative ptr for swift coroutines.

       A ptrauth intrinsic for swift co-routine support that allows creation of signed pointer
 from offset stored at address relative to the pointer.

 Following C-like pseudo code (ignoring keys,discriminators) explains its operation:
      let rawptr = PACauth(inputptr);
      return PACsign(  rawptr +  signextend64( *(int32*)(rawptr+addend) ))

 What: Authenticate a signed pointer, load a 32bit value at offset 'addend' from pointer,
       add this value to pointer, sign this new pointer.
 builtin: __builtin_ptrauth_auth_load_relative_and_sign
 intrinsic: ptrauth_auth_resign_load_relative

AbhayKanhere committed:
[CodeGen][AArch64] ptrauth intrinsic to safely construct relative pointer for swift coroutines
       A ptrauth intrinsic for swift co-routine support that allows creation of signed pointer
 from offset stored at address relative to the pointer.

 Following C-like pseudo code (ignoring keys,discriminators) explains its operation:
      let rawptr = PACauth(inputptr);
      return PACsign(  rawptr +  *(int32*)(rawptr+addend) )

 What: Authenticate a signed pointer, load a 32bit value at offset 'addend' from pointer,
       add this value to pointer, sign this new pointer.
 builtin: __builtin_ptrauth_auth_load_relative_and_sign
 intrinsic: ptrauth_auth_resign_load_relative

note: conflicts resolved llvm/lib/Target/AArch64/AArch64AsmPrinter.cpp due to upstream change
in args for emitPtrauthAuthResign

AbhayKanhere commented:
@ahmedbougacha  @jroelofs @rjmccall

rjmccall commented on clang/lib/Headers/ptrauth.h:
```diff
@@ -259,78 +284,6 @@ typedef __UINTPTR_TYPE__ ptrauth_generic_signature_t;
-
-#define ptrauth_cxx_vtable_pointer(key, address_discrimination,                \
-                                   extra_discrimination...)
-
```
These no-op forms were intentionally added so that clients who are willing to accept the non-portability of the existence of the header can simply call the intrinsic without extra `#if` checks at each use site. If there's a good reason to remove them, we can talk about it, but it should be done as a separate PR.

AbhayKanhere replied:
ok removed. thanks.I somehow missed notification on your comment.

AbhayKanhere replied:
done. added.

AbhayKanhere commented:
rebased with upstream changes - conflicts resolved. Added is passed in std::optional<uint64_t> to match how distinguish between AUTPAC and AUTPACRELLOAD.

rjmccall reviewed (changes requested):
Oliver is right that we need custom checking in SemaChecking akin to the other builtins, e.g. to validate that the keys are constants. The pattern should be straightforward to follow. Otherwise the frontend changes LGTM except as noted.

ojhunt commented:
> Oliver is right that we need custom checking in SemaChecking akin to the other builtins, e.g. to validate that the keys are constants. The pattern should be straightforward to follow. Otherwise the frontend changes LGTM except as noted.

When I noticed that I did wonder why we need custom type checking? we've typed the definition of the builtin and I recall a pile of very same-y custom checks in other builtins - is it just propagating types or is it something more complicated? (If there are some common idioms I'd be willing to spend some time adding support for them)

rjmccall commented on clang/lib/Sema/SemaChecking.cpp:
```diff
@@ -1814,6 +1814,33 @@ static ExprResult PointerAuthAuthAndResign(Sema &S, CallExpr *Call) {
+      AddendExpr->getIntegerConstantExpr(S.Context);
+  const Expr *Arg = Call->getArg(5)->IgnoreParenImpCasts();
+  DeclRefExpr *DRE = cast<DeclRefExpr>(Call->getCallee()->IgnoreParenCasts());
+  FunctionDecl *FDecl = cast<FunctionDecl>(DRE->getDecl());
```
Please put these inside the `if` block. Also, there's an `isIntegerConstantExpr` you can use if you don't care what the value is.

ojhunt commented on clang/lib/Headers/ptrauth.h:
```diff
@@ -380,7 +416,6 @@ typedef __UINTPTR_TYPE__ ptrauth_generic_signature_t;
     ((ptrauth_generic_signature_t)0);                                          \
   })
 
-
```
nit: revert :D

ojhunt commented on clang/lib/Headers/ptrauth.h:
```diff
@@ -178,6 +178,31 @@ typedef __UINTPTR_TYPE__ ptrauth_generic_signature_t;
+
+   Do not pass a null pointer to this function. A null pointer
+   will not successfully authenticate. */
+#define ptrauth_auth_load_relative_and_sign(__value, __old_key, __old_data,    \
```
I think I'd prefer this to be `__offset` rather than `__addend` unless `__addend` is the term used in swift, but even then it feels a bit off.

rjmccall replied:
No, I agree, "offset" is the user-friendly word.

ojhunt reviewed (approved):
I'm happy with the clang parts of this. Thanks.

It needs separate sign off on the llvm parts though.

rjmccall reviewed (approved):
I agree with Oliver's nitpicks, but otherwise, yeah, I'm happy with the clang changes.

AbhayKanhere committed:
renamed addend to offset, other nits

AbhayKanhere committed:
Updated glue/chain for intrinsic and a test.

AbhayKanhere committed:
resolved bad merge

AbhayKanhere committed:
return type mismatch is an error clang/test/Sema/ptrauth.c

AbhayKanhere commented:
@ahmedbougacha this PR is ready to merge.
